test-data/unit/check-type-typevar.test

```diff
@@ -0,0 +1,98 @@
+-- Test cases for type[SomeTypeVar] treatment
+
+
+[case testTypeOfTypeVarInstantiation]
+from typing import TypeVar, Type
+T = TypeVar('T')
+def f(x: Type[T]) -> T:
+    return x()
+
+[case testTypeOfTypeVarBoundInstantiationInitParams]
+from typing import TypeVar, Type
+class A:
+    def __init__(self, x: str) -> None: pass
+T = TypeVar('T', bound=A)
+def f(x: Type[T]) -> T:
+    return x('a')
+def f2(x: Type[T]) -> T:
+    return x(1)  # E: Argument 1 to "A" has incompatible type "int"; expected "str"
+
+[case testTypeOfTypeVarBoundInstantiationGeneric]
+from typing import Generic, TypeVar, Type
+class A:
+    def __init__(self, x: str) -> None: pass
+
+T = TypeVar('T', bound=object)
+V = TypeVar('V', bound=A)
+
+class Test(Generic[T, V]):
+    X: Type[T]
+    Y: Type[V]
+    def f1(self) -> T:
+        return self.X()
+    def f2(self) -> V:
+        return self.Y()  # E: Missing positional argument "x" in call to "A"
+    def f3(self) -> V:
+        return self.Y(1)  # E: Argument 1 to "A" has incompatible type "int"; expected "str"
+    def f4(self) -> V:
+        return self.Y('foo')
+
+[case testTypeOfTypeVarInstantiationInitParamsUnionBound]
+from typing import TypeVar, Type, Union
+class A:
+    def __init__(self, x: str) -> None: pass
+class B:
+    def __init__(self, x: int) -> None: pass
+class C:
+    def __init__(self, x: int) -> None: pass
+
+T = TypeVar('T', bound=Union[A, B])
+def f(x: Type[T]) -> T:
+    return x(1)  # E: Argument 1 to "A" has incompatible type "int"; expected "str"
+
+T2 = TypeVar('T2', bound=Union[B, C])
+def f2(x: Type[T2]) -> T2:
+    return x(1)
+
+[case testTypeOfTypeVarInstantiationVariants]
+from typing import TypeVar, Type
+class A:
+    def __init__(self, x: str) -> None: pass
+class B:
+    def __init__(self, x: int) -> None: pass
+class C:
+    def __init__(self, x: int) -> None: pass
+
+T = TypeVar('T', A, B)
+def f(x: Type[T]) -> T:
+    return x(1)  # E: Argument 1 to "A" has incompatible type "int"; expected "str"
+
+T2 = TypeVar('T2', B, C)
+def f2(x: Type[T2]) -> T2:
+    return x(1)
+
+def return_indirect(x: Type[T2]) -> T2:
```

Comment:
Why do we need indirect tests? 🤔

Reply:
Because `return some_immediate_expression_or_literal()` and `return some_var_with_already_fixed_type` are not 100% equivalent (consider `Literal` or `TypedDict` return, like [this](https://mypy-play.net/?mypy=master&python=3.11&flags=strict&gist=25664dc906a1b722a93d75366bd1032d)), and I'd like to be sure here. It may have no special treatment now (and it doesn't), but we may have a case in future where similar stricter inference is necessary.